### ISSUE TITLE
cache is located in site_packages

### ISSUE THREAD

**Issue by brunopostle** [IfcOpenShell/IfcOpenShell#5564]
The cache folder shouldn't really be a subfolder of the static data folder. Ideally it would be outside site_packages altogether, as site_packages ought to be read-only under normal operation.

ie. On this Linux system I would want this cache in `$HOME/.cache/bonsai` and on Windows it should be something like `C:\Users\<Username>\AppData\Local\bonsai\cache
`

Possibly the folder name should be versioned if there is a likelihood that the cache format may change in the future.

In addition, the user should be able to delete this cache folder at any time, so bonsai should check that it exists and create it recursively as needed.

**Comment by Andrej730** [comment]
But why is it a problem? We place "bonsai" in site-packages basically just as a hack to accommodate Blender requirements and keep it accessible as `bonsai` everywhere.

**Comment by Moult** [comment]
BTW I think we have a feature where you can customise the "data_dir" which includes everything and cache.

**Comment by brunopostle** [comment]
The cache folder is written to by the user, everything else here is distributed with the software.

Also on my system, all this stuff is in /usr/ so it is actually read-only. I'm bored of looking at cache errors, and wondering if this has a performance impact. 

I'll do a fix if nobody else sees this as a problem, though I dont know when.

**Comment by Moult** [comment]
I agree this is a problem. Also site packages may be controlled by Blender. It's not a super high priority problem though personally :) If you find the time to fix it, it'll be a great thing. I don't know how Python detects user app config folders like XDG_CONFIG_HOME and so on.

**Comment by Andrej730** [comment]
> Also site packages may be controlled by Blender.

Missed that point first time I read but it does makes the issue more critical.

Updating/disabling Bonsai always uninstalls all it's dependencies and when new version will be installed/Bonsai will be reenabled, `site-packages` is basically purged and reloaded from the wheels - possibly removing all modifications done by users. E.g. their pset templates library in `bonsai\bim\data\pset`, ifc templates, ifc libraries, sheets/titleblocks templates. Also possible assets (.css, .svg libraries) but it's more controversial.

So, I guess there is a need not for just `$HOME/.cache/bonsai` but `$HOME/bonsai` to store all this stuff.

**Comment by brunopostle** [comment]
On Linux this user specific template stuff should be in `$HOME/.local/share/bonsai`.

A cool feature would be if Bonsai looks for a file here first, then in a `share` folder in the Extension itself, then finally in `/usr/share/bonsai`.

**Comment by brunopostle** [comment]
I have a patch that makes this cache folder configurable, creates it if necessary, and defaults to a sane cross-platform location (ie. `~/.cache/bonsai` on linux).

It uses the [appdirs](https://pypi.org/project/appdirs/) module which will need to be bundled with bonsai.

appdirs is also the solution to the data_dir problem above, ie.

    appdirs.user_data_dir('bonsai', 'Bonsai')

..gives these locations depending on your platform:

    C:\Users\Bruno\AppData\Local\Bonsai\bonsai
    /Users/Bruno/Library/Application Support/bonsai
    /home/bruno/.local/share/bonsai

**Comment by brunopostle** [comment]
It looks like appdirs has been superceded by [platformdirs](https://pypi.org/project/platformdirs/)

**Comment by brunopostle** [comment]
Followup: I implemented this as a pull request (#5856) because it adds a `platformdirs` dependency, not because it breaks anything.

**Comment by Andrej730** [comment]
Let's also move 'data' folder outside Bonsai to fix those kind of issues - https://github.com/IfcOpenShell/IfcOpenShell/issues/5564#issuecomment-2487719642 ?

The current idea is this:

1) Set default value for BIMProperties.data_dir as `platformdirs.user_data_path("bonsai", roaming=True) / "data"`. 

`roaming` is a windows-only thing that will make it use `Roaming` instead of `Local`. No idea if users are actually using this (though Blender store it's settings in `Roaming`) but the idea of this folder intended by Microsoft that you can freely move configs and other stuff from it to other user profile and it will just work.

2) Add `.bonsai` text file to the `user_data_path` that will store current Bonsai version.

3) On `register` copy all files from `site-packages\bonsai\bim\data` to `user_data_path\data` if version doesn't match.

What do you think? @Moult @brunopostle

**Comment by Moult** [comment]
Go for it, having data in site packages is a terrible idea :)

Dion Moult

Sent from Proton Mail Android

-------- Original Message --------

**Comment by brunopostle** [comment]
The reason for putting this stuff into the user's home is so that they can customise it, if it is copied over on register then their customisations will get clobbered.

What would be ideal is if Bonsai looked for customisable files in this home folder, then looked in the current location if it can't find them. I think this would apply to just the assets and templates folders.

The Project Library and Property Set Templates pull-down lists could be constructed from files found in the current location and files found in libraries and pset folders in the users home. This would allow users to maintain their own, but also have them easily available in the UI.

There is some stuff that the user doesn't need to customise: fonts, gantt, icons, mpass, webui.

Having data mixed with python in site_packages isn't pretty, maybe this should just move to `libs/data` since there is lots of other non-python stuff in this libs folder.

**Comment by Andrej730** [comment]
Agreed, overriding makes more sense, it's a good idea. Then there also no need to worry about overriding custom user files.
